### PR TITLE
Verification with Standard JSON Input

### LINKED ISSUE #945: Solidity standard-json input verification
We should let users verify contracts using the standard-json input of the Solidity compiler.

Projects, particularly older ones, do not keep their metadata after compilation. These projects are unable to verify on Sourcify *directly*, they'd have to generate the metadata themselves. We can instead do this heavylifting for them and let more projects be verified.

Example: ENS contracts' solc-inputs: https://github.com/ensdomains/ens-contracts/blob/master/deployments/goerli/solcInputs/7948b60c3b601df824761a337a51d661.json

Closing the old issue about this topic: https://github.com/ethereum/sourcify/issues/319

### Session API

This is essentially a way to add files and not a new verification method.

Here the only change would be in how we add files. Instead of uploading a metadata.json necessarily, a user can provide a JSON-input file and a compiler version. Here we need to give a dropdown list of compiler versions, and an optional checkmark to include nightly builds. Because we need the additional compiler version, I'm suggesting a new endpoint instead of the existing `input-files`. We could parse to JSON input on the same endpoint but we'd need the version.

The server then feeds this to the respective compiler and generates the metadata files. If the metadata settings in the input JSON does not have `useLiteralContent: true`, we also need to generate the source files from the `content` of each source and add them to the session.

In the end the user will be left with multiple contracts with complete files but "chain & address missing". As usual, user can verify each contract by providing the chain and address.

`POST /session/input-solc-json`

Content-Type : `multipart/form-data` or `application/json`

If using `multipart/form-data`, the files should be in a field named `files` and compiler version under `compilerVersion`. If using `application/json`:

```
{
    "files": {
        "file-name1.json": ...,
        "file-name2.json": ...
    },
   "compilerVersion": "0.8.7+commit.e28d00a7"
}
```

### Non-session

This would be a new verification endpoint.

Here since potentially there will be multiple contracts but a single chain&address pair I suggest doing the following:

* Include an optional `chosenContract` field, as in `/verify`.
  * If it exists, verify the respective contract
    * TODO: Should this be an index as in `/verify` or a contract name, as it is in the standard-json itself?
* If not try verifying all contracts
  * TODO: Should we also try verifying all contracts in `/verify` if there are multiple contracts?

`POST /verify/solc-json`

Content-Type : multipart/form-data or application/json

```
{
    "address": ...,
    "chain": ...,
    "files": {
        "solc-input-1.json": ...,
        "solc-input-2.json": ...,
    },
    "compilerVersion": "0.8.7+commit.e28d00a7",
    // optional
    "chosenContract": 1, // index or name?
    // optional
    "contextVariables": {
      "abiEncodedConstructorArguments": ...,
      "msgSender": ...
    }
}
```

One thing to keep in mind is we should always add the relevant fields `metadata`, `deployedBytecode`, `immutableReferences` etc. to the `outputSelection` if they aren't added by the user.

## UI

Since this is a file input method, on the UI side we can add an "{} Import from Solidity JSON" button. When clicked this will show a dropdown list of compilers. Include a checkbox there that will "enable nightly builds".

Users can still drag and drop files or add by clicking. There the instructions should change to "Drag and drop the Solidity standard input-output JSON file" and after adding this will fetch the `/session/input-solc-json` endpoint, and not the `/input-files` as usual.

After getting the response from the server, populate the contracts, and the added files. Here the metadata files and the source files will be shown, as they are generated from the JSON file. The form should reset to its initial state.

## Todo's

* [x] Implementation
* [x] Tests
* [ ] Documentation

cc: @marcocastignoli appreciate a review on this

<sub>View in Huly <a href="https://app.huly.io/guest/w-kaanuzdogan-sourcify-65d30de0-75121c5c6c-fd615e?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NWU1N2U2YjhiNmVjZDA0ZDYyZDI4NDUiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6Incta2FhbnV6ZG9nYW4tc291cmNpZnktNjVkMzBkZTAtNzUxMjFjNWM2Yy1mZDYxNWUiLCJwcm9kdWN0SWQiOiIifQ.LTUH0L7zuI1OAl1s92eVIeKUnIZKikvHXoxpMjqVPxg">HI-416</a></sub>

### PR DESCRIPTION
**Pull request by kuzdogan**
Closes #945 

- Adds session and non session endpoints
- Adds tests
- Adds solc json verification to the UI:
  - Remove `CustomSelectSearch` and instead use the `SelectSearch` from the library itself with the correct types. CustomSelectSearch was there for a type workaround of the `onChange` prop: first removes the type, then adds the custom type. With the correct type from the library itself, there's no need for that.
  - Refactor `fuzzySearch` to its own module as it's reused. 
  - Refactor `ChainSelect` `GithubBranchSelect` 
  - Add the new `SolcJsonInput.tsx` with the out-of-the-box `SelectSearch`.

### CODE REVIEW

**Review by marcocastignoli — CHANGES_REQUESTED** [review]
Just two small things, great job! 🚀

File: src/server/controllers/VerificationController.ts [hunk 1]
```diff
@@ -196,6 +197,109 @@ export default class VerificationController
     res.send(getSessionJSON(session));
   };
 
+  private addInputSolcJsonEndpoint = async (req: Request, res: Response) => {
```
[the remark this answers is not in the document]
Comment (kuzdogan) [reply src/server/controllers/VerificationController.ts:200]: It's not actually a "verify" endpoint so I'd refrain from that.

Also other "file/contract add" endpoints are named similarly, that's why I named it that way
image

So either:
- Change all other endpoint names
- Keep all the same

**Comment by haltman-at** [comment]
Question from an interested outsider: What would happen if someone used this interface to attempt to verify a Yul contract, setting `"language": "Yul"` rather than `"language": "Solidity"`?

**Comment by kuzdogan** [comment]
@haltman-at Thanks for pointing this out! 

I tried with the examples in the Solidity docs [Yul page](https://docs.soliditylang.org/en/latest/yul.html#complete-erc20-example). The compiler compiles just fine but since the compiler output JSON does not have `.contracts`, no source code and metadata file get added to the session.

So we should check if the language is "Solidity"

**Review by marcocastignoli — APPROVED** [review]

**Pull request merged**